Catch saved tensors default hooks race condition

Varal7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61928
* #61927
* __->__ #61957
* #61834

If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

In the near future, we intend to make default hooks thread local so this bandaid should no longer be needed.

Differential Revision: [D29848525](https://our.internmc.facebook.com/intern/diff/D29848525)

Varal7 committed:
Catch saved tensors default hooks race condition

If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

[ghstack-poisoned]

Varal7 committed:
Update on "Catch saved tensors default hooks race condition"


If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

[ghstack-poisoned]

Varal7 committed:
Update on "Catch saved tensors default hooks race condition"


If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848525).

Varal7 committed:
Update on "Catch saved tensors default hooks race condition"


If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

Differential Revision: [D29848525](https://our.internmc.facebook.com/intern/diff/D29848525)

[ghstack-poisoned]

soulitzer reviewed (commented):
This looks fine to me. But instead of raising an error, could we just prevent the race condition entirely by using a lock?

Varal7 commented:
So the desired behavior is to raise and error. 
Under a lock:
Thread A takes a lock and sets default hooks.
Thread B waits for A to release lock.
Thread A releases lock.
Thread B takes a lock and sets default hooks: realizes that hooks are already set, throws an error.

albanD commented on torch/csrc/autograd/python_saved_variable_hooks.cpp:
```diff
@@ -46,16 +46,22 @@ namespace torch { namespace autograd {
-    unpack_hook_ = unpack_hook.release().ptr();
+    PyObject* pack_hook_ptr = pack_hook.release().ptr();
+    PyObject* unpack_hook_ptr = unpack_hook.release().ptr();
+    if (!pack_hook_.compare_exchange_strong(expected, pack_hook_ptr) ||
```
What happens if one succeed and one fails here?
Both threads would end up in the if statement and raise an error. And so this is ok?

Varal7 replied:
I expect that if any of the exchange fails, then the program will crash, raising an error. 
If the program is run again, most likely the TORCH_CHECK from line 55 will fail and indicate the user that something is wrong with his code (e.g. he is setting default hooks in the forward pass of a module that is passed to `torch.nn.DataParallel`).
We want to fail fast here and I expect this race condition to be so rare that if it does happen in practice, it's worth opening an issue?

zou3519 replied:
The mutex-based solution seems easier to reason about; with the mutex we could convince ourselves that if two threads are trying to set the global hook, exactly one of them will fail.

I'm having a hard time reasoning if there are problems with reference counts if two threads end up inside this conditional and call reset_hooks(). A user's program doesn't have to terminate if one of the threads raise an exception, right?

Varal7 replied:
Yes, my mistake. Now I've convinced myself that the current code is problematic. The mutex-based solution looks appealing.

Varal7 committed:
Update on "Catch saved tensors default hooks race condition"


If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

Differential Revision: [D29848525](https://our.internmc.facebook.com/intern/diff/D29848525)

[ghstack-poisoned]

zou3519 commented on torch/csrc/autograd/python_saved_variable_hooks.cpp:
```diff
@@ -46,16 +46,22 @@ namespace torch { namespace autograd {
 
   // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
-  PyObject* PyDefaultSavedVariableHooks::pack_hook_ = nullptr;
+  std::atomic<PyObject*> PyDefaultSavedVariableHooks::pack_hook_(nullptr);
```
There were some comments in #61834 about making the default hooks thread-local. Would that solve the problem of this race condition (and is that still the plan)?

Varal7 replied:
Yes, making the default hooks thread-local would solve the race condition. 
It is still the plan to make it thread local and I intend to start working on it within the next two weeks (my internship ends in 4)

zou3519 replied:
So this PR is just a stopgap solution for now?

Varal7 replied:
Yup.

Varal7 committed:
Update on "Catch saved tensors default hooks race condition"


If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

Differential Revision: [D29848525](https://our.internmc.facebook.com/intern/diff/D29848525)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848525).

Varal7 committed:
Update on "Catch saved tensors default hooks race condition"


If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

Differential Revision: [D29848525](https://our.internmc.facebook.com/intern/diff/D29848525)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848525).

Varal7 committed:
Update on "Catch saved tensors default hooks race condition"


If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

In the near future, we intend to make default hooks thread local so this bandaid should no longer be needed.

Differential Revision: [D29848525](https://our.internmc.facebook.com/intern/diff/D29848525)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848525).

zou3519 commented on torch/csrc/autograd/python_saved_variable_hooks.cpp:
```diff
@@ -46,16 +46,22 @@ namespace torch { namespace autograd {
+    if (!pack_hook_.compare_exchange_strong(expected, pack_hook_ptr) ||
+        !unpack_hook_.compare_exchange_strong(expected, unpack_hook_ptr)) {
+      reset_hooks();
+      TORCH_INTERNAL_ASSERT(false); // race condition
```
This should contain an error message so that the user is getting some context into what's happening

zou3519 reviewed (commented):
The mutex-based solution seems easier to reason about, but since this is a stopgap solution it doesn't really matter what we pick

Varal7 committed:
Update on "Catch saved tensors default hooks race condition"


If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

In the near future, we intend to make default hooks thread local so this bandaid should no longer be needed.

Differential Revision: [D29848525](https://our.internmc.facebook.com/intern/diff/D29848525)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848525).

Varal7 committed:
Update on "Catch saved tensors default hooks race condition"


If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

In the near future, we intend to make default hooks thread local so this bandaid should no longer be needed.

Differential Revision: [D29848525](https://our.internmc.facebook.com/intern/diff/D29848525)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848525).

zou3519 commented on torch/csrc/autograd/python_saved_variable_hooks.h:
```diff
@@ -28,10 +28,12 @@ struct PyDefaultSavedVariableHooks {
 
+  // Mutex to ensure that concurrent operations that modify default pack_hook_ and
+  // unpack_hook_ are thread-safe.
+  static std::mutex mutex_;
```
should this be in the Private section?

Varal7 committed:
Update on "Catch saved tensors default hooks race condition"


If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

In the near future, we intend to make default hooks thread local so this bandaid should no longer be needed.

Differential Revision: [D29848525](https://our.internmc.facebook.com/intern/diff/D29848525)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848525).

Varal7 committed:
Update on "Catch saved tensors default hooks race condition"


If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

In the near future, we intend to make default hooks thread local so this bandaid should no longer be needed.

Differential Revision: [D29848525](https://our.internmc.facebook.com/intern/diff/D29848525)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848525).
